core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/neo/DataProcessor.scala

```diff
@@ -0,0 +1,146 @@
+
+  private[this] def funToRunnable(fun: () => Unit): Runnable = new Runnable() { def run(): Unit = fun() }
+
+  // dp thread
+  private val executorService: ExecutorService = Executors.newSingleThreadExecutor
+
+  // dp thread stats:
+  private var inputTupleCount = 0L
+  private var outputTupleCount = 0L
+  private var currentInputTuple: Either[ITuple, InputExhausted] = _
+
+  // initialize dp thread upon construction
+  // we can use dpThreadFuture to kill dp thread if needed
+  private val dpThreadFuture: Future[_] = executorService.submit(funToRunnable(
+    () => {
+      try {
+        this.dpThread()
+      } catch {
+        case e: Exception =>
+          throw new RuntimeException(e)
+      }
+    }))
+
+
+  /** provide API for actor to get stats of this operator
+    * @return (input tuple count, output tuple count)
+    */
+  def collectStatistics():(Long,Long) = (inputTupleCount, outputTupleCount)
+
+  /** provide API for actor to get current input tuple of this operator
+    * @return current input tuple if it exists
+    */
+  def getCurrentInputTuple: ITuple = {
+    if(currentInputTuple != null && currentInputTuple.isLeft){
+      currentInputTuple.left.get
+    }else{
+      null
+    }
+  }
+
+  /** process currentInputTuple through operator logic.
+    * @return an iterator of output tuples
+    */
+  private[this] def consumeOneTuple(): Iterator[ITuple] = {
+    var outputIterator:Iterator[ITuple] = null
+    try{
+      outputIterator = operator.processTuple(currentInputTuple, tupleInput.getCurrentInput)
+      if(currentInputTuple.isLeft) inputTupleCount += 1
+    }catch{
+      case e:Exception =>
+        handleOperatorException(e, isInput = true)
+    }
+    outputIterator
+  }
+
+  /** transfer one tuple from iterator to downstream.
+    * @param outputIterator
+    */
+  private[this] def outputOneTuple(outputIterator:Iterator[ITuple]): Unit = {
+    var outputTuple: ITuple = null
+    try{
+      outputTuple = outputIterator.next
+    }catch{
+      case e:Exception =>
+        handleOperatorException(e, isInput = true)
+    }
+    if (outputTuple != null){
+      try{
+        outputTupleCount += 1
+        tupleOutput.transferTuple(outputTuple, outputTupleCount)
+      }catch{
+        case bp:BreakpointException =>
+          pauseUtil.pause(PauseUtil.Breakpoint)
+          self ! LocalBreakpointTriggered // TODO: apply FIFO & exactly-once protocol here
+        case e: Exception =>
+          handleOperatorException(e,isInput = false)
+      }
+    }
+  }
+
+  /** Provide main functionality of data processing
+    * @throws Exception (from engine code only)
+    */
+  @throws[Exception]
+  private[this] def dpThread(): Unit = {
+    // main DP loop: runs until all upstreams exhaust.
+    while (!tupleInput.isAllUpstreamsExhausted) {
+      // take the next input tuple from tupleInput, blocks if no tuple available.
+      currentInputTuple = tupleInput.nextInputTuple()
```

Comment:
Agreed. Also, there will be an issue if the worker is paused but the DP thread is blocked on waiting for input. If there is no input tuple, the DP thread will never reach pauseCheck so the DP thread will never be paused.

Reply:
Possible solutions:
1. If the worker paused, stash incoming input batches so the DP thread will be blocked on waiting for input until resume.
2. If the worker paused, add one dummy batch into the blocking queue to activate the pause check logic.
3. apply a timeout on blockingQueue.get

Reply:
Another solution:

use a variable to show if the DP thread is executing user's code, if yes, the worker actor will wait for the notification. If no, the actor can assume it is paused.

Reply:
I see another problem. Earlier DP thread wasn't using up resources if there was no input. But now, the thread is always there. This will cause even more problem if we try to implement migration after pausing. The DP thread will have to be killed I suppose?

Reply:
I finally decided to use insert a dummy batch into the blocking queue to unblock the DP thread because this solution will not introduce any race conditions or bugs.

Reply:
@shengquan-ni if you use the dummy batch approach, please not make it a subclass of Tuple or a special Tuple object. Similar to our InputExhausted design, please make it explicit by adding a separate type and user either.